File: src/ginger/hardware/adc.py
"""ADS7830 8-channel 8-bit ADC over I2C (address 0x48).

PCB V2.0: voltage coefficient 5.2, battery multiplier 2.
"""

import smbus2

_ADDRESS = 0x48
_COMMAND = 0x84  # single-ended, internal ref + ADC on

# Maps logical channel number to the ADS7830 MUX bits
_CHANNEL_MAP = {i: ((i << 2) | (i >> 1)) & 0x07 for i in range(8)}

_VOLTAGE_COEFF_V1 = 3.3
_VOLTAGE_COEFF_V2 = 5.2
_BATTERY_MULT_V1 = 3
_BATTERY_MULT_V2 = 2


class ADC:
    def __init__(self, pcb_version: int = 2, bus: int = 1):
        self._bus = smbus2.SMBus(bus)
        self._v_coeff = _VOLTAGE_COEFF_V2 if pcb_version == 2 else _VOLTAGE_COEFF_V1
        self._batt_mult = _BATTERY_MULT_V2 if pcb_version == 2 else _BATTERY_MULT_V1

    def read_raw(self, channel: int) -> int:
        """Return raw 8-bit ADC value for channel 0–7."""
        cmd = _COMMAND | (_CHANNEL_MAP[channel] << 4)
        self._bus.write_byte(_ADDRESS, cmd)
        # Read until two consecutive reads agree (noise filter)
        while True:
            v1 = self._bus.read_byte(_ADDRESS)
            v2 = self._bus.read_byte(_ADDRESS)
            if v1 == v2:
                return v1

    def read_voltage(self, channel: int) -> float:
        """Return voltage (V) for the given channel."""
        return round(self.read_raw(channel) / 255.0 * self._v_coeff, 2)

    def read_battery(self) -> float:
        """Return battery voltage (V). Wired to channel 2 with a voltage divider."""
        return round(self.read_voltage(2) * self._batt_mult, 2)

    def close(self) -> None:
        self._bus.close()

File: ginger-rs/src/adc.rs
//! ADS7830 8-channel 8-bit ADC over I2C (address 0x48).
//!
//! PCB V2.0: voltage coefficient 5.2, battery on channel 2 with ×2 divider.

use rppal::i2c::I2c;

use crate::Result;

const ADDRESS: u16 = 0x48;
const COMMAND: u8  = 0x84; // single-ended, internal ref + ADC on

const VOLTAGE_COEFF_V2: f32 = 5.2;
const BATTERY_MULT_V2:  f32 = 2.0;

// Maps channel 0–7 to ADS7830 MUX bits.
fn channel_cmd(ch: u8) -> u8 {
    let mux = ((ch << 2) | (ch >> 1)) & 0x07;
    COMMAND | (mux << 4)
}

pub struct Adc {
    i2c: I2c,
    v_coeff:   f32,
    batt_mult: f32,
}

impl Adc {
    pub fn new() -> Result<Self> {
        let mut i2c = I2c::new()?;
        i2c.set_slave_address(ADDRESS)?;
        Ok(Self { i2c, v_coeff: VOLTAGE_COEFF_V2, batt_mult: BATTERY_MULT_V2 })
    }

    /// Raw 8-bit ADC value for channel 0–7 (reads until two consecutive reads agree).
    ///
    /// ADS7830 uses raw byte I2C (no register address), so we use send_byte / receive_byte.
    pub fn read_raw(&mut self, channel: u8) -> Result<u8> {
        self.i2c.smbus_send_byte(channel_cmd(channel))?;
        loop {
            let v1 = self.i2c.smbus_receive_byte()?;
            let v2 = self.i2c.smbus_receive_byte()?;
            if v1 == v2 {
                return Ok(v1);
            }
        }
    }

    pub fn read_voltage(&mut self, channel: u8) -> Result<f32> {
        let raw = self.read_raw(channel)? as f32;
        Ok((raw / 255.0 * self.v_coeff * 100.0).round() / 100.0)
    }

    /// Battery voltage (V). Wired to channel 2 with a voltage divider.
    pub fn read_battery(&mut self) -> Result<f32> {
        let v = self.read_voltage(2)?;
        Ok((v * self.batt_mult * 100.0).round() / 100.0)
    }

    /// Left ambient light sensor voltage (V). Higher = brighter.
    pub fn read_light_left(&mut self) -> Result<f32> {
        self.read_voltage(0)
    }

    /// Right ambient light sensor voltage (V). Higher = brighter.
    pub fn read_light_right(&mut self) -> Result<f32> {
        self.read_voltage(1)
    }
}
